lib/pyinotify.py

```diff
@@ -0,0 +1,2346 @@
+    operations.
+    """
+    def __init__(self, msg, wmd):
+        """
+        @param msg: Exception string's description.
+        @type msg: string
+        @param wmd: This dictionary contains the wd assigned to paths of the
+                    same call for which watches were successfully added.
+        @type wmd: dict
+        """
+        self.wmd = wmd
+        Exception.__init__(self, msg)
+
+
+class WatchManager:
+    """
+    Provide operations for watching files and directories. Its internal
+    dictionary is used to reference watched items. When used inside
+    threaded code, one must instanciate as many WatchManager instances as
+    there are ThreadedNotifier instances.
+
+    """
+    def __init__(self, exclude_filter=lambda path: False):
+        """
+        Initialization: init inotify, init watch manager dictionary.
+        Raise OSError if initialization fails, raise InotifyBindingNotFoundError
+        if no inotify binding was found (through ctypes or from direct access to
+        syscalls).
+
+        @param exclude_filter: boolean function, returns True if current
+                               path must be excluded from being watched.
+                               Convenient for providing a common exclusion
+                               filter for every call to add_watch.
+        @type exclude_filter: callable object
+        """
+        self._ignore_events = False
+        self._exclude_filter = exclude_filter
+        self._wmd = {}  # watch dict key: watch descriptor, value: watch
+
+        self._inotify_wrapper = INotifyWrapper.create()
+        if self._inotify_wrapper is None:
+            raise InotifyBindingNotFoundError()
+
+        self._fd = self._inotify_wrapper.inotify_init() # file descriptor
+        if self._fd < 0:
+            err = 'Cannot initialize new instance of inotify, %s'
+            raise OSError(err % self._inotify_wrapper.str_errno())
+
+    def close(self):
+        """
+        Close inotify's file descriptor, this action will also automatically
+        remove (i.e. stop watching) all its associated watch descriptors.
+        After a call to this method the WatchManager's instance become useless
+        and cannot be reused, a new instance must then be instanciated. It
+        makes sense to call this method in few situations for instance if
+        several independant WatchManager must be instanciated or if all watches
+        must be removed and no other watches need to be added.
+        """
+        os.close(self._fd)
+
+    def get_fd(self):
+        """
+        Return assigned inotify's file descriptor.
+
+        @return: File descriptor.
+        @rtype: int
+        """
+        return self._fd
+
+    def get_watch(self, wd):
+        """
+        Get watch from provided watch descriptor wd.
+
+        @param wd: Watch descriptor.
+        @type wd: int
+        """
+        return self._wmd.get(wd)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

You are using the `get` method without a default argument to return the value of a key in a dictionary. We recommended that you use a default argument so that if the value for your key is not found, a default value is returned. If a default value is not provided and the key is not found, then `None` is returned.

[Learn more](https://docs.python.org/3/library/stdtypes.html#dict.get)